url8df97b3310_dbathon_nurikabe_java_tgz-pJ8-dbathon_nurikabe_MainBenchJ8/src/dbathon/nurikabe/board/Board.java

```diff
@@ -0,0 +1,327 @@
+    return cells[coordsToIndex(x, y)];
+  }
+
+  public int getCellCount() {
+    return width * height;
+  }
+
+  public int getSolutionWhiteCount() {
+    return solutionWhiteCount;
+  }
+
+  public int getSolutionBlackCount() {
+    return getCellCount() - getSolutionWhiteCount();
+  }
+
+  public Stream<Cell> getCells() {
+    return cellsListView.stream();
+  }
+
+  public Stream<Cell> getWhiteCells() {
+    return getCells().filter(cell -> cell.isWhite());
+  }
+
+  public Stream<Cell> getBlackCells() {
+    return getCells().filter(cell -> cell.isBlack());
+  }
+
+  public Stream<Cell> getUnknownCells() {
+    return getCells().filter(cell -> cell.isUnknown());
+  }
+
+  public int getCount(CellColor cellColor) {
+    return (int) getCells().filter(cell -> cell.getColor() == cellColor).count();
+  }
+
+  public int getMaxNumber() {
+    return maxNumber;
+  }
+
+  public Set<Cell> getNeighborsSet(Cell cell) {
+    if (cell.getBoard() != this) {
+      throw new IllegalArgumentException("cell");
+    }
+
+    return neighborSets.get(coordsToIndex(cell.getX(), cell.getY()));
+  }
+
+  public Stream<Cell> getNeighbors(Cell cell) {
+    return getNeighborsSet(cell).stream();
+  }
+
+  private void connectWhiteCell(Cell cell) {
+    final FixedCell fixedCell = cell.getFixedCell();
+    if (fixedCell != null) {
+      getNeighbors(cell).filter(neighbor -> neighbor.isWhite() && neighbor.getFixedCell() == null)
+          .forEach(neighbor -> {
+            neighbor.setFixedCell(fixedCell);
+            connectWhiteCell(neighbor);
+          });
+    }
+  }
+
+  public void connectWhiteCells() {
+    getWhiteCells().forEach(this::connectWhiteCell);
+  }
+
+  public Map<FixedCell, Set<Cell>> getWhiteGroupsWithFixedCell() {
+    connectWhiteCells();
+
+    return getCells().filter(cell -> cell.getFixedCell() != null)
+        .collect(
+            Collectors.groupingBy(Cell::getFixedCell,
+                Collectors.toCollection(() -> new CellSet(this))));
+  }
+
+  /**
+   * Recursively search all connected cells.
+   */
+  private void findConnectedCells(Cell cell, Set<Cell> result) {
+    result.add(cell);
+    getNeighbors(cell).filter(neighbor -> neighbor.getColor() == cell.getColor())
+        .filter(neighbor -> !result.contains(neighbor))
+        .forEach(neighbor -> findConnectedCells(neighbor, result));
+  }
+
+  public Set<Set<Cell>> getGroups(CellColor cellColor) {
+    final Set<Cell> cellsWithColor = new CellSet(this);
+    for (final Cell cell : cells) {
+      if (cell.getColor() == cellColor) {
+        cellsWithColor.add(cell);
+      }
+    }
+
+    final Set<Set<Cell>> result = new HashSet<>();
+    while (!cellsWithColor.isEmpty()) {
+      final Set<Cell> group = new CellSet(this);
+      final Cell startCell = cellsWithColor.iterator().next();
+      findConnectedCells(startCell, group);
+      result.add(group);
+
+      cellsWithColor.removeAll(group);
+    }
+
+    return result;
+  }
+
+  public boolean isSolution() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 231-239, lines 247-257 into separate methods.